kernelci/api/__init__.py

```diff
@@ -31,9 +31,120 @@ def __init__(self, config: kernelci.config.api.API, token: str):
+
+    @property
+    def headers(self) -> dict:
+        """HTTP headers with content type, authorization token etc."""
+        return self._headers
+
+
+class Base:
+    """Common primitive methods used in API bindings implementation"""
+
+    def __init__(self, data: Data):
+        self._data = data
+
+    @property
+    def data(self) -> Data:
+        """Internal Data object instance"""
+        return self._data
+
+    def make_url(self, path: str) -> str:
+        """Make a full URL for a given API endpoint path"""
+        version_path = '/'.join((self.data.config.version, path))
+        return urllib.parse.urljoin(self.data.config.url, version_path)
+
+    def _get(self, path, params=None):
+        url = self.make_url(path)
+        resp = requests.get(
+            url, params, headers=self.data.headers,
+            timeout=self.data.timeout
+        )
+        resp.raise_for_status()
+        return resp
+
+    def _post(self, path, data=None, params=None, json_data=True):
+        url = self.make_url(path)
+        if json_data:
+            jdata = json.dumps(data)
+            resp = requests.post(
+                url, jdata, headers=self.data.headers,
+                params=params, timeout=self.data.timeout
+            )
+        else:
+            headers = self.data.headers.copy()
+            headers['Content-Type'] = 'application/x-www-form-urlencoded'
+            resp = requests.post(
+                url, data, headers=headers,
+                params=params, timeout=self.data.timeout
+            )
+        resp.raise_for_status()
+        return resp
+
+    def _put(self, path, data=None, params=None):
+        url = self.make_url(path)
+        jdata = json.dumps(data)
+        resp = requests.put(
+            url, jdata, headers=self.data.headers,
+            params=params, timeout=self.data.timeout
+        )
+        resp.raise_for_status()
+        return resp
+
+    def _patch(self, path, data=None, params=None):
+        url = self.make_url(path)
+        jdata = json.dumps(data)
+        resp = requests.patch(
+            url, jdata, headers=self.data.headers,
+            params=params, timeout=self.data.timeout
+        )
+        resp.raise_for_status()
+        return resp
+
+    def _get_paginated(self, input_params, path, offset=None, limit=None):
+        params = input_params.copy()
+
+        if any((offset, limit)):
+            params.update({
+                'offset': offset or None,
+                'limit': limit or None,
+            })
+            return self._get(path, params=params)
+
+        objs = []
+        offset = 0
+        limit = 100
+        params['limit'] = limit
+        while True:
+            params['offset'] = offset
+            resp = self._get(path, params=params)
+            items = resp.json()['items']
+            objs.extend(items)
+            if len(items) < limit:
+                break
+            offset += limit
+        return objs
```

Comment:
Shouldn't this be in the `API` class?

Reply:
No, because then nested classes can't access it.  That's why I created a separate Base class.  Then the API class as well as the nested ones for separate parts of the API can all inherit from it.